leases/models.py

```diff
@@ -0,0 +1,126 @@
+from django.core.exceptions import ValidationError
+from django.db import models
+from django.utils.translation import ugettext_lazy as _
+from enumfields import EnumField
+
+from customers.models import Boat, CustomerProfile
+from reservations.models import BerthReservation, WinterStorageReservation
+from resources.models import Berth, WinterStoragePlace
+from utils.models import TimeStampedModel, UUIDModel
+
+from .enums import LeaseStatus
+
+
+class AbstractLease(TimeStampedModel, UUIDModel):
+    customer = models.ForeignKey(
+        CustomerProfile, verbose_name=_("customer"), on_delete=models.PROTECT
+    )
+    boat = models.ForeignKey(
+        Boat, verbose_name=_("customer's boat"), on_delete=models.PROTECT
+    )
+
+    status = EnumField(
+        LeaseStatus,
+        verbose_name=_("lease status"),
+        max_length=30,
+        default=LeaseStatus.OFFERED,
+    )
+
+    start_date = models.DateField(verbose_name=_("start date"))
+    end_date = models.DateField(verbose_name=_("end date"))
+
+    comment = models.TextField(verbose_name=_("comment"), blank=True)
+
+    class Meta:
+        abstract = True
+
+    def clean(self):
+        if not self.boat.owner == self.customer:
+            raise ValidationError(
+                _("The boat should belong to the customer who is creating the lease")
+            )
+
+    def save(self, *args, **kwargs):
+        # ensure full_clean is always ran
+        self.full_clean()
+
+        # this relies on the related_name="changes" for both
+        # BerthLeaseChange.lease and WinterStorageLeaseChange.lease
+        if not self._state.adding:
+            original = type(self).objects.get(pk=self.pk)
+            if original.status != self.status:
+                self.changes.create(from_status=original.status, to_status=self.status)
+        super().save(*args, **kwargs)
+
+
+class BerthLease(AbstractLease):
+    berth = models.ForeignKey(
+        Berth, verbose_name=_("berth"), on_delete=models.PROTECT, related_name="leases"
+    )
+    application = models.ForeignKey(
+        BerthReservation,
+        verbose_name=_("application"),
+        on_delete=models.SET_NULL,
+        blank=True,
+        null=True,
+        related_name="leases",
+    )
+
+    def __str__(self):
+        return " {} > {} - {} ({})".format(
+            self.berth, self.start_date, self.end_date, self.status
+        )
+
+
+class WinterStorageLease(AbstractLease):
+    place = models.ForeignKey(
+        WinterStoragePlace,
+        verbose_name=_("place"),
+        on_delete=models.PROTECT,
+        related_name="leases",
+    )
+    application = models.ForeignKey(
+        WinterStorageReservation,
+        verbose_name=_("application"),
+        on_delete=models.SET_NULL,
+        blank=True,
+        null=True,
+        related_name="leases",
+    )
+
+    def __str__(self):
+        return " {} > {} - {} ({})".format(
+            self.place, self.start_date, self.end_date, self.status
+        )
+
```

Comment:
I think you could use [`default_related_name`](https://docs.djangoproject.com/en/3.0/ref/models/options/#default-related-name) for this model and `BerthLease` if you want to have proper related names for customer and boat.

Reply:
oh, good idea, looks pretty neat! thanks 🙂